File: lib/services/executors/claude-agent/src/cli-env.ts
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.apache at the repo root.

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

export interface CliAuthConfig {
  anthropicApiKey: string;
  claudeCodeOauthToken: string;
}

export interface CliEnvResult {
  env: Record<string, string>;
  cleanup: () => void;
}

export function buildCliEnv(config: CliAuthConfig): CliEnvResult {
  const basePath = process.env.PATH ?? "/usr/local/bin:/usr/bin:/bin";

  if (config.anthropicApiKey) {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "claude-agent-env-"));
    const keyFilePath = path.join(tmpDir, "api-key");
    fs.writeFileSync(keyFilePath, config.anthropicApiKey, { mode: 0o600 });

    const helperPath = path.join(tmpDir, "api-key-helper.sh");
    fs.writeFileSync(helperPath, `#!/bin/sh\ncat '${keyFilePath}'`, { mode: 0o700 });

    const claudeDir = path.join(tmpDir, ".claude");
    fs.mkdirSync(claudeDir);
    fs.writeFileSync(
      path.join(claudeDir, "settings.json"),
      JSON.stringify({ apiKeyHelper: helperPath }),
    );

    return {
      env: { HOME: tmpDir, PATH: basePath },
      cleanup: () => {
        try { fs.unlinkSync(keyFilePath); } catch {}
        try { fs.unlinkSync(helperPath); } catch {}
        try { fs.unlinkSync(path.join(claudeDir, "settings.json")); } catch {}
        try { fs.rmdirSync(claudeDir); } catch {}
        try { fs.rmdirSync(tmpDir); } catch {}
      },
    };
  }

  const env: Record<string, string> = {
    HOME: process.env.HOME ?? os.homedir(),
    PATH: basePath,
  };
  if (config.claudeCodeOauthToken) {
    env.CLAUDE_CODE_OAUTH_TOKEN = config.claudeCodeOauthToken;
  }
  return { env, cleanup: () => {} };
}

File: lib/services/executors/claude-agent/clienv.go
// Copyright © 2026 Fall Guy Consulting.
// Dual-licensed under AGPL-3.0-or-later or a Fall Guy Consulting commercial
// license. See LICENSE.agpl and COPYRIGHT at the repo root.

package claudeagent

import (
	"encoding/json"
	"os"
	"path/filepath"
)

type CliAuthConfig struct {
	AnthropicAPIKey      string
	ClaudeCodeOauthToken string
}

type CliEnvResult struct {
	Env     map[string]string
	Cleanup func()
}

func BuildCliEnv(config CliAuthConfig) (CliEnvResult, error) {
	basePath := os.Getenv("PATH")
	if basePath == "" {
		basePath = "/usr/local/bin:/usr/bin:/bin"
	}

	if config.AnthropicAPIKey != "" {
		tmpDir, err := os.MkdirTemp("", "claude-agent-env-")
		if err != nil {
			return CliEnvResult{}, err
		}
		succeeded := false
		defer func() {
			if !succeeded {
				_ = os.RemoveAll(tmpDir)
			}
		}()
		keyFilePath := filepath.Join(tmpDir, "api-key")
		if err := os.WriteFile(keyFilePath, []byte(config.AnthropicAPIKey), 0o600); err != nil {
			return CliEnvResult{}, err
		}
		helperPath := filepath.Join(tmpDir, "api-key-helper.sh")
		helper := "#!/bin/sh\ncat '" + keyFilePath + "'"
		if err := os.WriteFile(helperPath, []byte(helper), 0o700); err != nil {
			return CliEnvResult{}, err
		}
		claudeDir := filepath.Join(tmpDir, ".claude")
		if err := os.Mkdir(claudeDir, 0o755); err != nil {
			return CliEnvResult{}, err
		}
		settings, err := json.Marshal(map[string]string{"apiKeyHelper": helperPath})
		if err != nil {
			return CliEnvResult{}, err
		}
		if err := os.WriteFile(filepath.Join(claudeDir, "settings.json"), settings, 0o644); err != nil {
			return CliEnvResult{}, err
		}
		succeeded = true
		return CliEnvResult{
			Env: map[string]string{"HOME": tmpDir, "PATH": basePath},
			Cleanup: func() {
				_ = os.RemoveAll(tmpDir)
			},
		}, nil
	}

	home := os.Getenv("HOME")
	if home == "" {
		home, _ = os.UserHomeDir()
	}
	env := map[string]string{"HOME": home, "PATH": basePath}
	if config.ClaudeCodeOauthToken != "" {
		env["CLAUDE_CODE_OAUTH_TOKEN"] = config.ClaudeCodeOauthToken
	}
	return CliEnvResult{Env: env, Cleanup: func() {}}, nil
}
